### PR TITLE
Add GitHubReleasePlugin with createGitHubRelease task

### PR DESCRIPTION
**Pull request by sjohnr**
Closes gh-10457

### CODE REVIEW

File: buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleasePlugin.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	@Override
+	public void apply(Project project) {
+		project.getTasks().register("createGitHubRelease", CreateGitHubReleaseTask.class, new Action<CreateGitHubReleaseTask>() {
+			@Override
+			public void execute(CreateGitHubReleaseTask createGitHubRelease) {
+				createGitHubRelease.setGroup("Release");
+				createGitHubRelease.setDescription("Create a github release");
+				createGitHubRelease.dependsOn("generateChangelog");
+
+				createGitHubRelease.setGitHubAccessToken((String) project.findProperty("gitHubAccessToken"));
```
Comment (eleftherias) [inline buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleasePlugin.java:36]: I wonder if we should check if `gitHubAccessToken` exists before using it, similar to `GitHubMilestonePlugin`.

File: buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleaseApi.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	 */
+	public void publishRelease(RepositoryRef repository, Release release) {
+		String url = this.baseUrl + "/repos/" + repository.getOwner() + "/" + repository.getName() + "/releases";
+		String json = this.gson.toJson(release);
+		RequestBody body = RequestBody.create(MediaType.parse("application/json"), json);
+		Request request = new Request.Builder().url(url).post(body).build();
+		try {
+			Response response = this.httpClient.newCall(request).execute();
+			if (!response.isSuccessful()) {
+				throw new RuntimeException("Could not create release " + release + ". Got response " + response);
```
Comment (eleftherias) [inline buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleaseApi.java:64]: It may be more readable to use `release.getName()` in the exception messages instead of printing the entire object.

File: buildSrc/src/main/java/org/springframework/gradle/github/release/CreateGitHubReleaseTask.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+				.name(this.version)
+				.body(body)
+				.preRelease(false)
+				.build();
+		github.publishRelease(this.repository, release);
+	}
+
+	private String readReleaseNotes() {
+		Project project = getProject();
+		File inputFile = project.file(Paths.get(project.getBuildDir().getPath(), "changelog/release-notes.md"));
```
Comment (eleftherias) [inline buildSrc/src/main/java/org/springframework/gradle/github/release/CreateGitHubReleaseTask.java:58]: Rather than hardcoding "changelog/release-notes.md" here, we can have a public constant in `GitHubChangelogPlugin`, for example `public static final String RELEASE_NOTES_PATH = "changelog/release-notes.md";` and access it here as `GitHubChangelogPlugin.RELEASE_NOTES_PATH`. (the variable name could be anything)

File: buildSrc/src/main/java/org/springframework/gradle/github/release/RepositoryRef.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.gradle.github.release;
+
+/**
+ * Steve Riesenberg
+ */
+public class RepositoryRef {
```
Comment (eleftherias) [inline buildSrc/src/main/java/org/springframework/gradle/github/release/RepositoryRef.java:22]: This looks like it's the same class as `org.springframework.gradle.github.milestones.RepositoryRef`. Could we extract it into `org.springframework.gradle.github` and reuse it in both `milestones` and `release` packages?

**Review by eleftherias — COMMENTED** [review]
Looks good @sjohnr! I have left some suggestions inline.

File: buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleasePlugin.java (hunk 1 above)
Comment (sjohnr) [reply buildSrc/src/main/java/org/springframework/gradle/github/release/GitHubReleasePlugin.java:36]: It looks like the GitHubMilestonePlugin does not need an access token to work, since it's accessing public data. I will explore the right way to fail with a meaningful error message in this plugin, since it requires access token.

**Comment by sjohnr** [comment]
@eleftherias, I have updated the PR, including adding additional options `createRelease` and `branch`, and made `gitHubAccessToken` optional with a check, so that we can test with dry runs. This task also tags the release, so it accomplishes both #10456 and #10457.

I've tested it and it's working well, so we should be able to begin work on automating releases in a pipeline. Let me know if you spot anything else.

File: buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestonePlugin.java [hunk 5]
```diff
@@ -29,7 +29,7 @@ public void execute(GitHubMilestoneHasNoOpenIssuesTask githubCheckMilestoneHasNo
 				githubCheckMilestoneHasNoOpenIssues.setGroup("Release");
 				githubCheckMilestoneHasNoOpenIssues.setDescription("Checks if there are any open issues for the specified repository and milestone");
 				githubCheckMilestoneHasNoOpenIssues.setMilestoneTitle((String) project.findProperty("nextVersion"));
-				if (project.hasProperty("githubAccessToken")) {
```
Comment (eleftherias) [inline buildSrc/src/main/java/org/springframework/gradle/github/milestones/GitHubMilestonePlugin.java:32]: I think this should be a separate commit since it's independent from adding the new plugin.

**Review by eleftherias — CHANGES_REQUESTED** [review]
I left one comment @sjohnr. In my option you can merge this after that change. I don't think another review is needed.
